### ISSUE TITLE
Trial should fail test methods which are generator functions

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3917]
|<img alt="terrycojones's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| terrycojones reported|
|-|-|
|Trac ID|trac#3917|
|Type|enhancement|
|Created|2009-07-16 22:29:26Z|
|Branch|https://github.com/twisted/twisted/tree/check-for-generator-functions-3917|

I just added a test to an existing test suite and ran it. It passed, but I was uncomfortable because it seemed to complete too quickly. I changed an assert to its opposite and the test still passed. Hmm...

I realized I was yielding deferreds in the test, but had omitted the @defer.inlineCallbacks decorator. So when trial called my test, it got back a generator and thought it was done.

I think it would be nice if trial detected this and either logged a warning or passed the generator to defer._inlineCallbacks. While doing the latter might be considered overkill, a warning is at least in order.

<details><summary>Searchable metadata</summary>

```
trac-id__3917 3917
type__enhancement enhancement
reporter__terrycojones terrycojones
priority__normal normal
milestone__ 
branch__branches_check_for_generator_functions_3917 branches/check-for-generator-functions-3917
branch_author__habnabit habnabit
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1247783366000000 1247783366000000
changetime__1395164919000000 1395164919000000
version__None None
owner__habnabit habnabit
cc__terrycojones cc__exarkun cc__jerith
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This scenario is one reason to do TDD - when you expect the test to fail the first time you run it, you know right away that you did something wrong when it passes, instead. :)

One use case to be careful of is one where a test method is implemented in terms of a second test method.  The second test method might return a value which the first test method needs in order to run its test.  Presently, this is harmless, since trial ignores the return value when it is not a Deferred.  This change could introduce spurious warnings for that case.

One possibility would be to allow this warning to be suppressed, which wouldn't be terribly onerous, I think, as this is not a very common thing to do.

Another possibility might be to inspect the test method to determine if it is a generator function.  I don't think this would have similar issues with spurious warnings, because I can't think of any cases where one would want a test method to be a generator function (this is distinct from checking if a test method returns a generator, as a non-generator function may return a generator).

**Comment by twisted-trac** [comment]
|<img alt="terrycojones's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|terrycojones commented|
|-|-|

Thanks for the comments - I hadn't thought of any of that :-)  Typical.

Re TDD, I agree, but we're still left with the case where someone inadvertently (cut & paste, quite likely) adds a yield to a test case without adding inlineCallbacks. Then the test passes forevermore. I guess you might add that if you add code to a test, you should always do it in a way that makes the test fail.

Maybe I just need to practice more rigorous self discipline...

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_4"></a>|@habnabit commented|
|-|-|

Checking if test methods are generator functions is relatively straightforward—there's even `inspect.isgeneratorfunction`, which _does_ also check methods, in addition to functions. This should be a fairly small change, especially since Twisted depends on python 2.6+ now.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit set owner to @habnabit|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_6"></a>|@habnabit commented|
|-|-|

(In [40268]) Branching to 'check-for-generator-functions-3917'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_7"></a>|@habnabit commented|
|-|-|

[Buildbot](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/check-for-generator-functions-3917) looks fine, except for some false positives on twistedchecker and pyflakes. The new errors listed aren't from the code for this branch. There's also an error from a win32 builder which seems to intermittently fail, and the error ("No space left on device") doesn't appear to be relevant.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_10"></a>|@habnabit commented|
|-|-|

(In [40275]) Fail on generator function test methods instead of warn.

refs [#3917](https://github.com/twisted/twisted/issues/3917)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_11"></a>|@habnabit commented|
|-|-|

(In [40276]) Update the topfile too.

refs [#3917](https://github.com/twisted/twisted/issues/3917)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_12"></a>|@habnabit commented|
|-|-|

(In [40277]) Restore python 3 compatibility; fix pyflakes errors.

refs [#3917](https://github.com/twisted/twisted/issues/3917)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_13"></a>|@habnabit commented|
|-|-|

This should be ready for review again now. I've changed it to fail tests instead of issue a warning, which is much easier to catch.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_14"></a>|@habnabit commented|
|-|-|

Oh, right—[buildbot](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/check-for-generator-functions-3917) got rerun, and still has false positives on pyflakes/twistedchecker and accrued another failing builder which looks unrelated.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_15"></a>|@habnabit commented|
|-|-|

(In [40344]) Applying patch from icchasethi. refs [#3917](https://github.com/twisted/twisted/issues/3917)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_16"></a>|@habnabit commented|
|-|-|

(In [40346]) Revert "Applying patch from icchasethi. refs [#3917](https://github.com/twisted/twisted/issues/3917)"

This reverts commit 8982b5197e17d1844633828541905e00b2618f90.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_17"></a>|@habnabit commented|
|-|-|

The above was a mistake and should be properly reverted.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)|@Julian set owner to @Julian|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_19"></a>|@Julian commented|
|-|-|

Hey! Cool.

This feature makes me feel a bit uneasy tbh, but I get the underlying sentiment that there's a confusing class of errors that this could help.

Couple of minor review points:

 * You missed 2 spots when rejiggering things, the test case name and docstring for the tests still say Warnings when they're now failures.
 * Please use assertTrue instead of assert_, or here just self.fail, even though these lines aren't meant to be run.
 * As implemented, this feature will obviously only work for trial test cases. Would it be reasonable to attempt to implement this as a feature of the runner instead, so that this works for all test case instances?
 * I think the message is going to be a bit confusing for people who don't necessarily put 2 and 2 together as to why a test method being a generator is inherently a failure. I don't have any great ideas, but at the very least a loud(er) warning in some prose docs or a docstring somewhere telling people that there is this guard in place might be helpful in being a place they might find if they can't figure out why their test is failing. Or maybe adding the words "will never run" or something of an explanation to the message would be possible without making it too verbose.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)|@Julian set owner to @habnabit|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_21"></a>|@habnabit commented|
|-|-|

In order:

 1. Fixed.
 2. Fixed to use `self.fail`.
 3. I've just been looking through the code for the test runner, and it seems like putting this where I put it is the _easiest_ way to implement it. I'll keep looking, but I'm not sure putting it in the runner is feasible. Either way, this is "intended" to prevent against accidentally forgetting to apply `@defer.inlineCallbacks`, which requires trial test cases anyway.
 4. Reworded the failure message slightly. I think this is a lot clearer now. I could add a FAQ entry to the wiki about this too.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_22"></a>|@habnabit commented|
|-|-|

(In [41922]) Address review feedback. refs [#3917](https://github.com/twisted/twisted/issues/3917)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_23"></a>|@habnabit commented|
|-|-|

(In [41923]) Clarify the topfile. refs [#3917](https://github.com/twisted/twisted/issues/3917)

**Comment by twisted-trac** [comment]
|[<img alt="jerith's avatar" src="https://avatars.githubusercontent.com/u/43731?s=50" width="50" height="50">](https://github.com/jerith)<a name="note_24"></a>|@jerith commented|
|-|-|

Should this be an ERROR rather than a FAIL, perhaps? FAILs should come from failed assertions rather than incorrect test construction.

Apart from that, this looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_25"></a>|@habnabit commented|
|-|-|

(In [41926]) Report generator function test methods as errors, not failures.

refs [#3917](https://github.com/twisted/twisted/issues/3917)

**Comment by twisted-trac** [comment]
|[<img alt="jerith's avatar" src="https://avatars.githubusercontent.com/u/43731?s=50" width="50" height="50">](https://github.com/jerith)<a name="note_26"></a>|@jerith commented|
|-|-|

The test names and topfile are wrong again after this change. They should mention "error" rather than "failure". Perhaps the topfile should also explain that this is to catch missing @inlineCallbacks decorators?

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_27"></a>|@habnabit commented|
|-|-|

(In [41935]) Address review comments. refs [#3917](https://github.com/twisted/twisted/issues/3917)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_28"></a>|@habnabit commented|
|-|-|

[Buildbot looks good](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/check-for-generator-functions-3917), except for a pyflakes false positive, the Windows 7 builder being a bit slow, and the Windows XP builder being flaky. The XP builder in particular appears to be failing the same way in other branches, so I don't think it's related to this change. I'm rerunning it for the latest change right now, but it's small and I don't expect there will be any more failures.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_29"></a>|@habnabit removed owner|
|-|-|

still waiting on Windows 7, but I'll put this up for review anyway.

**Comment by twisted-trac** [comment]
|[<img alt="jerith's avatar" src="https://avatars.githubusercontent.com/u/43731?s=50" width="50" height="50">](https://github.com/jerith)<a name="note_30"></a>|@jerith set owner to @habnabit|
|-|-|

The new pyflakes error is just a line number change, so I'm not concerned about that. It may be worth changing the import to "from twisted.python.failure import Failure" to avoid masking it in all the functions with parameters named "failure", but I think that's out of scope for this ticket. The Windows build failures appear to be unrelated to these changes and I seem to recall Windows builds being flakey in general anyway.

I think this is good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_31"></a>|@habnabit set status to `closed`|
|-|-|

(In [41938]) Merge check-for-generator-functions-3917: Report test methods which are generator functions in trial TestCases as errors.

Author: habnabit
Reviewers: Julian, jerith
Fixes: [#3917](https://github.com/twisted/twisted/issues/3917)

Previously, if a trial TestCase had a test method which was a
generator function, the test would appear to pass even though its
code hadn't run at all. This behavior sometimes occurred if a
test was intended to be decorated with defer.inlineCallbacks, but
the test writer had forgotten to apply the decorator.
